grafana/synthetic-monitoring-app
# Plugin datasource RBAC

Will close: https://github.com/grafana/synthetic-monitoring-app/issues/877

## Problem

Currently the Synthetic Monitoring application works on the dated concept of user roles to work out whether a user has access to view / create / edit content. These roles are no longer the source of truth since Grafana introduced RBAC and despite the default access control settings for each user type still map correctly, administrators can now give fine-grained control and it is no longer accurate if these roles accurately reflect what permissions a user should have.

Synthetic Monitoring relies on three datasources to operate fully but it is still possible to use the Synthetic Monitoring product in some capacity if you only have access to one of the datasources and should reflect accordingly.

We have had several issues and escalations related to this problem and it is also on our roadmap to support RBAC fully at an app-level so this puts us in an excellent position when that work starts.

## Solution

_Please note this solution is purely for the plugin. The backend api for the Synthetic Monitoring datasource does not support RBAC yet._

This PR removes all references and functions which relate to checking a user's role. It creates a `PermissionsContext` which is a child of the `SMDatasourceProvider` and makes the associated call to each datasource the user has available with the `accessControl=true` parameter to establish what permissions a user has per datasource.

It is possible that a user may have view and/or edit permissions to the Synthetic Monitoring plugin but does not have permissions whatsoever to know what metrics and logs datasource it is connected to. It is also possible the inverse is true and that a user can view metrics and logs but not have access to the Synthetic Monitoring datasource.

### Understanding Grafana RBAC

Grafana RBAC has multiple levels which are important to understand. There are permissions that a user has at a global level and there are permissions that users have at a datasource level.

There are currently two areas of our application where we need to know what permissions a user has at the **global Grafana** level:

1. `datasources:create` -- this permission is needed to create the initial Synthetic Monitoring Datasource. It is provided to the admin role by default but not to editor role.
2. `plugins:write` -- this permission is needed for a user to enable / disabled the Synthetic Monitoring plugin. It is provided to the admin role by default but not to editor role.

There is a `useDSPermission` file and it provides multiple utility hooks which abstract the PermissionsContext and are now used throughout the application to work out what it should display to a user: 

```
useDSPermission
```
This is the underlying hook that powers the following hooks. We do use this hook directly in one place: the populate alerts button on the alerts page as we need to know if they have the permission `alert.instances.external:write`.

```
useCanReadSM()
useCanWriteSM()
```
These are used extensively and inform much of the behaviour of what we show and when we disable creation / editing / deleting Checks and Probes across the application.

```
useCanReadMetrics()
useCanWriteMetrics() 
```
In almost all cases we do not need to know if a user can write metrics. This hook is only used once to inform whether we should link to the metrics datasource on the config page. However, it is important that we know if a user can read metrics. If they cannot they are not shown any of the metrics on the Probes / Checks pages.

```
useCanReadLogs()
useCanWriteLogs()
```
Similar to above with needing to know about if a user can write logs. However there are only two places we currently need to know whether a user has logs access or not:

1. The test button for checks creation (and soon to be the test button in requests)
2. The logs panel in the Scenes dashboards


> [!NOTE]
> **About Scenes dashboards:** I have done very little work on hiding panels in Scenes based on user permissions. It is very hard to work with and I'll come back to this in a future PR. There is a **lot** of work that needs to be done on the scenes dashboards on a whole so I want to more informed opinion before tackling this properly. 

### Screenshots of different permission combinations

**No read access to Synthetic Monitoring AND no permission to create datasources**. The user doesn't know whatsoever if Synthetic Monitoring is set up or not but they do have access to view the plugin.

<img width="1412" alt="image" src="https://github.com/user-attachments/assets/5f17d8b9-f06f-4948-a510-48dae575b928">

<img width="1428" alt="image" src="https://github.com/user-attachments/assets/dd692546-876f-4925-a88e-9cbbf2b2ecb5">

<img width="1429" alt="image" src="https://github.com/user-attachments/assets/499e4c13-e6d9-4bfd-8126-5ff084f520c4">

**No write access to Synthetic Monitoring**

<img width="1409" alt="image" src="https://github.com/user-attachments/assets/88d954e9-20cb-4247-801e-a3be94e0c21c">

<img width="1416" alt="image" src="https://github.com/user-attachments/assets/bc20a6d4-f780-4ffb-85e6-e96738ffee02">

<img width="1416" alt="image" src="https://github.com/user-attachments/assets/6209087c-e292-45f2-9383-cb4bf7549dcd">

<img width="1412" alt="image" src="https://github.com/user-attachments/assets/cae82368-a270-440a-be60-8af2c48ec38c">


**No read access to the Metrics datasource**

<img width="1416" alt="image" src="https://github.com/user-attachments/assets/7ed74467-76c3-40fc-9b92-49a663581020">

<img width="1419" alt="image" src="https://github.com/user-attachments/assets/3bc31495-dfeb-4470-baf2-e63e6c1fb4dd">

<img width="1417" alt="image" src="https://github.com/user-attachments/assets/c9156e57-c05d-404a-9424-39d679be3634">

<img width="1428" alt="image" src="https://github.com/user-attachments/assets/220afb3b-ec7f-4f2e-95a7-71e58987f981">

<img width="1431" alt="image" src="https://github.com/user-attachments/assets/766bca71-1db4-4e64-b08b-fe85309960ba">

**No write access to the Metrics datasource**

<img width="1425" alt="image" src="https://github.com/user-attachments/assets/a63af82b-8d95-4387-a549-1bd3eb62fc28">

**No read access to the logs datasource**

<img width="1424" alt="image" src="https://github.com/user-attachments/assets/0ab1cf24-2ebd-4dbe-abac-efc88bde650d">

<img width="1429" alt="image" src="https://github.com/user-attachments/assets/5abc577a-15f1-48cb-a997-36265db85ccf">

## How to test

I have made myself a matrix of users locally with different permissions and I have been logging in and out of them checking if what permissions I have granted each user match up with what they should be allowed to do. The users I have made are:

Admin (default role admin)
Editor (default role editor)
Viewer (default role editor)

For each datasource remove the default role permissions and only grant access to them on a per user basis.

NoSmEditor (role editor with no SM datasource access)
OnlyLogsEditor (role editor with edit access to SM + Logs but no metrics access)
OnlyMetricsEditor (role editor with edit access to SM + Metrics but no logs access)

NoSMViewer (role viewer with no SM datasource access)
OnlyLogsViewer (role viewer with view access to SM + Logs but no metrics access)
OnlyMetricsViewer (role viewer with view access to SM + Metrics but no logs access)

I have also added integration tests as appropriate.
